### DISCUSSION TITLE
A new implementation of rsMatrix

### DISCUSSION THREAD

**Discussion by RobinSchmidt** [RobinSchmidt/RS-MET#317] [Show and tell]
my old class rsMatrix was implemented before C++11 and i created some complicated copy-on-write mechanism in order to avoid extra memory allocations for the functions that return a matrix - in particular, the arithmetic operators +,-,*. due to the fact that in C++ pre 11, function return values were returned by-value, a naive implementation of an operator, say +, would have to create a local variable of type matrix (one heap allocation) and then to return it, create a copy (another heap allocation). ...and maybe even a third in assignment? so in a statement like `C = A + B`, the operator would have two allocations and the assignment a third one? ...not sure. in any case, there would be some extra allocations. to avoid these extra allocations, i implemented this copy-on-write stuff - which is messy

nowadays, however, C++ supports rvalue references and return-value optimization - so all this messy stuff is not necessary anymore. all i have to do now is to carefully implement implement the copy- and move-constructors and copy- and move assignment operators - which is what i did with the new implementation.

the new class is - for the time being - called `rsMatrixNew`. the old one shall be phased out soon and then the new one will get the proper name rsMatrix, when the transition is complete.

**Comment by RobinSchmidt** [comment]
it's actually a shame that there is nothing like std::matrix in the STL. i mean, matrices are such a fundamental data-structure for all numerical work. who cares about unordered_multimaps? :-P ...rant over

**Comment by elanhickler** [comment]
pfft, stupid unordered_multimaps.

**Comment by RobinSchmidt** [comment]
ha! swapping two matrices naively causes 3 heap memory allocations whereas swapping them via `std::move` causes none! i changed the implementation of `RAPT::rsSwap` to use move semantics. this will benefit any swapping operation of large(ish) data-structures - such as matrices but also std::vector, std::string, etc.

![image](https://user-images.githubusercontent.com/1833099/68674182-a8c40880-0555-11ea-8e0d-6c24d76d39ac.png)

![image](https://user-images.githubusercontent.com/1833099/68674224-c2655000-0555-11ea-9339-feaa0227d090.png)

**Comment by RobinSchmidt** [comment]
oh - i guess, i have just re-invented `std::swap`:
https://en.cppreference.com/w/cpp/algorithm/swap

**Comment by elanhickler** [comment]
for  RAPT::rsSpectrogramProcessor.getComplexSpectrogram, seems like you have rows and columns swapped.

Also, why isn't this working? I'm trying to feather/fade the amplitude above and below a row. The results seem to be on or off, no fading takes place.

```
Matrix mat = sp.getComplexSpectrogram(origAudio.getReadPointer(ch), samples);

...inside a loop...

mat(matC, harmonicProfile[c][r] - 4) *= 3.0/4.0;
mat(matC, harmonicProfile[c][r] - 3) *= 2.0/4.0;
mat(matC, harmonicProfile[c][r] - 2) *= 1.0/4.0;
mat(matC, harmonicProfile[c][r] - 1) *= 0.0/4.0;
mat(matC, harmonicProfile[c][r]    ) *= 0.0/4.0;
mat(matC, harmonicProfile[c][r] + 1) *= 0.0/4.0;
mat(matC, harmonicProfile[c][r] + 2) *= 1.0/4.0;
mat(matC, harmonicProfile[c][r] + 3) *= 2.0/4.0;
mat(matC, harmonicProfile[c][r] + 4) *= 3.0/4.0;
```

**Comment by RobinSchmidt** [comment]
oh - i missed this. is this still a problem? i just added a unit test for the +=, *=,... operators operating on elements and it seems to work:

```cpp
bool testMatrixOperators()
{
  bool ok = true;  // test result

  //.....

  // test element-wise +=, *=, -=, /=
  double tmp;
  tmp = A(2,1); A(2,1) += 2.0; ok &= A(2,1) == tmp + 2.0;
  tmp = A(2,1); A(2,1) *= 2.0; ok &= A(2,1) == tmp * 2.0;
  tmp = A(2,1); A(2,1) -= 2.0; ok &= A(2,1) == tmp - 2.0;
  tmp = A(2,1); A(2,1) /= 2.0; ok &= A(2,1) == tmp / 2.0;

  return ok;
}
```

**Comment by elanhickler** [comment]
go ahead and close the issue

**Comment by RobinSchmidt** [comment]
this is actually meant to be a discussion thread (i.e. not primarily a bug-thread), so it should remain open. i may move it into the discussions section soon, if github has fixed this feature meanwhile

**Comment by RobinSchmidt** [comment]
Yay! We finally get std::mdspan and std::mdarray

https://www.youtube.com/watch?v=aFCLmQEkPUw

...so, that's basically...**tensors**! With rank = 2, this can be used for matrices as well. I guess that will become the new way of doing these things. It's similar in purpose to this:

https://github.com/RobinSchmidt/RS-MET/blob/master/Libraries/RobsJuceModules/rapt/Data/MultiArray.h

but needs the rank (i.e. number of dimensions) to be known at compile time. That's probably a good trade-off decision. In rsMultiArray, it can be changed at runtime - at the price of having to store the extents and strides in a std::vector of int.
